### ISSUE TITLE
how can i split batch and num_boxes after ROIAlign?

### ISSUE THREAD

**Issue by waterbroz** [matterport/Mask_RCNN#732]
thanks for your sharing. It is really excellent work, and give me help a lot.
I am doubt that how can i split batch and num_boxes after ROIAlign? In author's code '    box_to_level = tf.concat(box_to_level, axis=0)
    box_range = tf.expand_dims(tf.range(tf.shape(box_to_level)[0]), 1)
    box_to_level = tf.concat([tf.cast(box_to_level, tf.int32), box_range],
                             axis=1)
    # Rearrange pooled features to match the order of the original boxes
    # Sort box_to_level by batch then box index
    # TF doesn't have a way to sort by two columns, so merge them and sort.
    sorting_tensor = box_to_level[:, 0] * 100000 + box_to_level[:, 1]
    ix = tf.nn.top_k(sorting_tensor, k=tf.shape(box_to_level)[0]).indices[::-1]
    ix = tf.gather(box_to_level[:, 2], ix)
    pooled = tf.gather(pooled, ix)
    # Re-add the batch dimension
    pooled = tf.expand_dims(pooled, 0)
    return pooled'

i think shape returned should be(1,batchsize*num_boxes,poolsize,poolsize,c) not (batchsize,num_boxes,poolsize,poolsize,c). Am I get something mistake? thanks

**Comment by wmalara** [comment]
I think you're right that the last expand_dims doesn't create the correct output shape. I've noticed strange results when using predict with batch size greater than 1: the output shape of mrcnn_mask  was [batch, batch*num_boxes, height, width, channels] instead of [batch, num_boxes, height, width, channels]. 
Replacing
`pooled = tf.expand_dims(pooled, 0)`
with
`boxes_shape = tf.shape(boxes)`
`pooled = tf.reshape(pooled, [boxes_shape[0], boxes_shape[1], self.pool_shape[0], self.pool_shape[1], tf.shape(feature_maps[0])[-1]])`
helped, but I'm not sure if this is the most elegant way to split the first dim into batches.

**Comment by DevinCheung** [comment]
@waterbroz how did you solve the problem? thx~

**Comment by waterbroz** [comment]
@DevinCheung   `    sorting_tensor = box_to_level[:, 0] * 100000 + box_to_level[:, 1]
    ix = tf.nn.top_k(sorting_tensor, k=tf.shape(
        box_to_level)[0]).indices[::-1]
    ix = tf.gather(box_to_level[:, 2], ix)
    results = []
    for i in range(batch_size):
        results.append(tf.gather(pooled, ix[i*num_boxes:(i+1)*num_boxes]))

    results = tf.stack(results, axis=0)
    return results`

**Comment by wmalara** [comment]
I think this issue has been fixed in #919.
